File: lib/features/war_cwl/models/war_cwl.dart
import 'package:clashkingapp/features/war_cwl/models/cwl_league.dart';
import 'package:clashkingapp/features/war_cwl/models/cwl_league_round.dart';
import 'package:clashkingapp/features/war_cwl/models/war_info.dart';
import 'package:clashkingapp/features/war_cwl/models/war_member_presence.dart';
import 'package:clashkingapp/core/utils/debug_utils.dart';

class WarCwl {
  final String tag;
  final bool isInWar;
  final bool isInCwl;
  final WarInfo warInfo;
  final CwlLeague? leagueInfo;
  final List<WarInfo> warLeagueInfos;

  WarCwl({
    required this.tag,
    required this.isInWar,
    required this.isInCwl,
    required this.warInfo,
    this.leagueInfo,
    required this.warLeagueInfos,
  });

  int get teamSize => warLeagueInfos.isNotEmpty
      ? warLeagueInfos[0].teamSize ?? 0
      : warInfo.teamSize ?? 0;

  factory WarCwl.fromJson(Map<String, dynamic> json, String? tag) {
    try {
      final warInfoData = _asMap(json['war_info']);
      final warState = warInfoData?['state']?.toString() ?? 'notInWar';
      final leagueInfoData = _asMap(json['league_info']);
      final warLeagueInfosData =
          (json['war_league_infos'] as List<dynamic>? ?? [])
              .map(_asMap)
              .whereType<Map<String, dynamic>>()
              .map(WarInfo.fromJson)
              .where((warInfo) =>
                  warInfo.state != 'unknown' ||
                  warInfo.clan != null ||
                  warInfo.opponent != null)
              .toList();

      return WarCwl(
        tag: json['clan_tag']?.toString() ?? tag ?? '',
        isInWar: json['isInWar'] == true,
        isInCwl: json['isInCwl'] == true,
        warInfo: warState != 'notInWar'
            ? WarInfo.fromJson(_asMap(warInfoData?['currentWarInfo']))
            : WarInfo(state: 'notInWar'),
        leagueInfo:
            leagueInfoData != null ? CwlLeague.fromJson(leagueInfoData) : null,
        warLeagueInfos: warLeagueInfosData,
      );
    } catch (e) {
      DebugUtils.debugError(" Error parsing WarCwl: $e");
      return WarCwl(
        tag: json['clan_tag']?.toString() ?? tag ?? '',
        isInWar: false,
        isInCwl: false,
        warInfo: WarInfo(state: 'unknown'),
        leagueInfo: null,
        warLeagueInfos: [],
      );
    }
  }

  WarInfo? getActiveWarByTag(String tag) {
    try {
      // Normalize clan tag for comparison (ensure # prefix)
      String normalizeClanTag(String clanTag) {
        if (!clanTag.startsWith('#')) return '#$clanTag';
        return clanTag;
      }

      final normalizedTag = normalizeClanTag(tag);
      DebugUtils.debugInfo("🔍 Looking for CWL war for clan: $normalizedTag");

      bool matchesTag(WarInfo warInfo, String tag) =>
          (warInfo.clan != null &&
              normalizeClanTag(warInfo.clan!.tag) == tag) ||
          (warInfo.opponent != null &&
              normalizeClanTag(warInfo.opponent!.tag) == tag);

      // First try to find active war (inWar)
      final activeWar = warLeagueInfos.firstWhere(
        (warInfo) =>
            warInfo.state == 'inWar' && matchesTag(warInfo, normalizedTag),
        orElse: () => WarInfo(state: 'notFound'),
      );

      if (activeWar.state != 'notFound') {
        DebugUtils.debugSuccess("Found active CWL war for clan $tag");
        return activeWar.reorderForClan(normalizedTag);
      }

      // If no active war, try preparation war
      final prepWar = warLeagueInfos.firstWhere(
        (warInfo) =>
            warInfo.state == 'preparation' &&
            matchesTag(warInfo, normalizedTag),
        orElse: () => WarInfo(state: 'notFound'),
      );

      if (prepWar.state != 'notFound') {
        DebugUtils.debugSuccess(
            "Found preparation CWL war for clan $normalizedTag");
        return prepWar.reorderForClan(normalizedTag);
      }

      // If no active or prep war, try most recent ended war
      final endedWars = warLeagueInfos
          .where(
            (warInfo) =>
                warInfo.state == 'warEnded' &&
                matchesTag(warInfo, normalizedTag),
          )
          .toList();

      if (endedWars.isNotEmpty) {
        // Sort by end time to get most recent
        endedWars.sort((a, b) => (b.endTime ?? '')
            .toString()
            .compareTo((a.endTime ?? '').toString()));
        DebugUtils.debugSuccess(
            "Found recent ended CWL war for clan $normalizedTag");
        return endedWars.first.reorderForClan(normalizedTag);
      }

      DebugUtils.debugError("No CWL wars found for clan $normalizedTag");
      DebugUtils.debugInfo("🔍 Available wars in warLeagueInfos:");
      for (final war in warLeagueInfos) {
        DebugUtils.debugInfo(
            "   War: ${war.state}, Clan: ${war.clan?.tag}, Opponent: ${war.opponent?.tag}");
      }
      return null;
    } catch (e) {
      DebugUtils.debugError("Error finding CWL war for clan $tag: $e");
      return null;
    }
  }

  CwlLeagueRound? getRoundForWarTag(String? warTag) {
    final league = leagueInfo;
    if (league == null) {
      return CwlLeagueRound(roundNumber: -1, warTags: []);
    }

    return league.rounds.firstWhere(
      (round) => round.containsWar(warTag),
      orElse: () => CwlLeagueRound(roundNumber: -1, warTags: []),
    );
  }

  WarInfo? getWarInfoFromTag(String tag) {
    return warLeagueInfos.firstWhere(
      (warInfo) => warInfo.tag == tag,
      orElse: () => WarInfo(state: 'unknown'),
    );
  }

  WarInfo getActiveWarForClan(String clanTag) {
    // Use the improved logic from getActiveWarByTag
    final war = getActiveWarByTag(clanTag);
    return war ?? WarInfo(state: 'notInCwl');
  }

  WarMemberPresence getMemberPresence(String memberTag, String clanTag) {
    try {
      final warInfo = getActiveWarForClan(clanTag);
      final member = warInfo.getMemberByTag(memberTag);

      if (member != null) {
        int attacksDone = member.attacks?.length ?? 0;
        return WarMemberPresence(
          isInWar: true,
          attacksDone: attacksDone,
          attacksAvailable: warInfo.attacksPerMember ?? 1,
        );
      }
    } catch (e) {
      DebugUtils.debugError("Error getting member presence: $e");
    }
    return WarMemberPresence(isInWar: false);
  }

  /// Get active war by player tag and automatically reorder so user's clan is always 'clan'
  /// This is a convenience method that combines finding the war and reordering for the user
  WarInfo? getActiveWarByPlayerTag(String playerTag) {
    try {
      // First find any war containing this player
      for (final warInfo in warLeagueInfos) {
        final playerInClan =
            warInfo.clan?.members.any((member) => member.tag == playerTag) ??
                false;
        final playerInOpponent = warInfo.opponent?.members
                .any((member) => member.tag == playerTag) ??
            false;

        if (playerInClan || playerInOpponent) {
          // Found a war with this player, prioritize by state
          if (warInfo.state == 'inWar') {
            DebugUtils.debugSuccess(
                "Found active CWL war for player $playerTag");
            return warInfo.reorderForUser(playerTag);
          }
        }
      }

      // If no active war, try preparation
      for (final warInfo in warLeagueInfos) {
        final playerInClan =
            warInfo.clan?.members.any((member) => member.tag == playerTag) ??
                false;
        final playerInOpponent = warInfo.opponent?.members
                .any((member) => member.tag == playerTag) ??
            false;

        if (playerInClan || playerInOpponent) {
          if (warInfo.state == 'preparation') {
            DebugUtils.debugSuccess(
                "Found preparation CWL war for player $playerTag");
            return warInfo.reorderForUser(playerTag);
          }
        }
      }

      // If no active or prep war, try most recent ended war
      final playerWars = warLeagueInfos.where((warInfo) {
        final playerInClan =
            warInfo.clan?.members.any((member) => member.tag == playerTag) ??
                false;
        final playerInOpponent = warInfo.opponent?.members
                .any((member) => member.tag == playerTag) ??
            false;
        return (playerInClan || playerInOpponent) &&
            warInfo.state == 'warEnded';
      }).toList();

      if (playerWars.isNotEmpty) {
        // Sort by end time to get most recent
        playerWars.sort((a, b) => (b.endTime ?? '')
            .toString()
            .compareTo((a.endTime ?? '').toString()));
        DebugUtils.debugSuccess(
            "Found recent ended CWL war for player $playerTag");
        return playerWars.first.reorderForUser(playerTag);
      }

      DebugUtils.debugWarning("No CWL wars found for player $playerTag");
      return null;
    } catch (e) {
      DebugUtils.debugError("Error finding CWL war for player $playerTag: $e");
      return null;
    }
  }
}

Map<String, dynamic>? _asMap(dynamic value) {
  if (value is Map<String, dynamic>) {
    return value;
  }
  if (value is Map) {
    return Map<String, dynamic>.from(value);
  }
  return null;
}

File: expo/src/features/war/models/war-cwl.ts
import { CwlLeague, CwlLeagueRound } from './cwl';
import { normalizeWarTag, record, records, string, type JsonRecord } from './parsing';
import { WarInfo, WarMemberPresence } from './war';

export class WarCwl {
  constructor(
    readonly tag: string,
    readonly isInWar: boolean,
    readonly isInCwl: boolean,
    readonly warInfo: WarInfo,
    readonly leagueInfo: CwlLeague | null,
    readonly warLeagueInfos: readonly WarInfo[],
  ) {}

  static fromJson(json: JsonRecord, fallbackTag: string | null = null): WarCwl {
    try {
      const warInfoData = record(json.war_info);
      const warState = string(warInfoData.state, 'notInWar');
      const leagueInfoData = record(json.league_info);
      const warLeagueInfos = records(json.war_league_infos)
        .map(WarInfo.fromJson)
        .filter((war) => war.state !== 'unknown' || war.clan !== null || war.opponent !== null);
      return new WarCwl(
        string(json.clan_tag, fallbackTag ?? ''),
        json.isInWar === true,
        json.isInCwl === true,
        warState !== 'notInWar'
          ? WarInfo.fromJson(warInfoData.currentWarInfo)
          : new WarInfo('notInWar'),
        Object.keys(leagueInfoData).length ? CwlLeague.fromJson(leagueInfoData) : null,
        warLeagueInfos,
      );
    } catch {
      return new WarCwl(
        string(json.clan_tag, fallbackTag ?? ''),
        false,
        false,
        new WarInfo('unknown'),
        null,
        [],
      );
    }
  }

  get teamSize(): number {
    for (const war of this.warLeagueInfos) if ((war.teamSize ?? 0) > 0) return war.teamSize!;
    if ((this.warInfo.teamSize ?? 0) > 0) return this.warInfo.teamSize!;
    for (const war of [...this.warLeagueInfos, this.warInfo]) {
      const size = lineupSizeForClan(war, this.tag);
      if (size > 0) return size;
    }
    return 0;
  }

  getActiveWarByTag(tag: string): WarInfo | null {
    try {
      const normalized = ensureHashPrefix(tag);
      const matches = (war: WarInfo) =>
        (war.clan !== null && ensureHashPrefix(war.clan.tag) === normalized) ||
        (war.opponent !== null && ensureHashPrefix(war.opponent.tag) === normalized);
      for (const state of ['inWar', 'preparation'] as const) {
        const found = this.warLeagueInfos.find((war) => war.state === state && matches(war));
        if (found) return found.reorderForClan(normalized);
      }
      const ended = this.warLeagueInfos
        .filter((war) => war.state === 'warEnded' && matches(war))
        .sort((left, right) =>
          (right.endTime?.toISOString() ?? '').localeCompare(left.endTime?.toISOString() ?? ''),
        );
      return ended[0]?.reorderForClan(normalized) ?? null;
    } catch {
      return null;
    }
  }

  getRoundForWarTag(warTag: string | null): CwlLeagueRound {
    return (
      this.leagueInfo?.rounds.find((round) => round.containsWar(warTag)) ??
      new CwlLeagueRound(-1, [])
    );
  }

  getWarInfoFromTag(tag: string): WarInfo {
    return this.warLeagueInfos.find((war) => war.tag === tag) ?? new WarInfo('unknown');
  }

  getActiveWarForClan(clanTag: string): WarInfo {
    return this.getActiveWarByTag(clanTag) ?? new WarInfo('notInCwl');
  }

  getMemberPresence(memberTag: string, clanTag: string): WarMemberPresence {
    try {
      const war = this.getActiveWarForClan(clanTag);
      if (war.state !== 'inWar') return new WarMemberPresence(false);
      const member = war.getMemberByTag(memberTag);
      return member
        ? new WarMemberPresence(true, member.attacks?.length ?? 0, war.effectiveAttacksPerMember)
        : new WarMemberPresence(false);
    } catch {
      return new WarMemberPresence(false);
    }
  }

  getActiveWarByPlayerTag(playerTag: string): WarInfo | null {
    try {
      const contains = (war: WarInfo) =>
        war.clan?.members.some((member) => member.tag === playerTag) === true ||
        war.opponent?.members.some((member) => member.tag === playerTag) === true;
      for (const state of ['inWar', 'preparation'] as const) {
        const found = this.warLeagueInfos.find((war) => war.state === state && contains(war));
        if (found) return found.reorderForUser(playerTag);
      }
      const ended = this.warLeagueInfos
        .filter((war) => war.state === 'warEnded' && contains(war))
        .sort((left, right) =>
          (right.endTime?.toISOString() ?? '').localeCompare(left.endTime?.toISOString() ?? ''),
        );
      return ended[0]?.reorderForUser(playerTag) ?? null;
    } catch {
      return null;
    }
  }
}

function lineupSizeForClan(war: WarInfo, clanTag: string): number {
  const normalized = normalizeWarTag(clanTag);
  if (normalizeWarTag(war.clan?.tag) === normalized) return war.clan?.members.length ?? 0;
  if (normalizeWarTag(war.opponent?.tag) === normalized) return war.opponent?.members.length ?? 0;
  return 0;
}

function ensureHashPrefix(tag: string): string {
  return tag.startsWith('#') ? tag : `#${tag}`;
}
